### PR TITLE
AwsSdk2Transport throw exception when using ApacheHttpClient to make an unsupported DELETE/GET request with a body

### LINKED ISSUE #503: [Maintenance] Unit test coverage on AwsSdk2Transport
### What is the gap?
Paths in AwsSdk2Transport should be covered by unit tests that are more light weighted than integration tests.

### LINKED ISSUE #521: [BUG] Delete PIT and clear scroll request fails with sigv4 signature mismatch for java client 2.5.0
### What is the bug?
`deletePit` request to opensearch client fails with sigv4 signing mismatch

```
org.opensearch.client.opensearch._types.OpenSearchException: Request failed: [security_exception] The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.

The Canonical String for this request should have been
'DELETE
/_search/point_in_time

... remaining request info...
```

### How can one reproduce the bug?
Create an OpenSearchClient following instructions from here (https://opensearch.org/docs/2.7/clients/java/#connecting-to-amazon-opensearch-service) and then attempt to use the client's `deletePit` API.

I believe this is because it is a  DELETE request with a body. The `createPit` method works as intended. Sending the request results in a sigv4 error

### What is the expected behavior?
The `deletePit` API call does not have a signature mismatch

### What is your host/environment?
Mac OS

```
implementation 'org.opensearch.client:opensearch-java:2.5.0'
implementation 'org.opensearch.client:opensearch-rest-client:2.7.0'
```

### Do you have any screenshots?
_If applicable, add screenshots to help explain your problem._

### Do you have any additional context?
_Add any other context about the problem._

### LINKED ISSUE #712: [BUG] Clear scroll request fails with sigv4 signature mismatch for java client 2.8.0
### What is the bug?
`clearScroll` request to opensearch client fails with sigv4 signing mismatch:

org.opensearch.client.opensearch._types.OpenSearchException: Request failed: [security_exception] The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.

The Canonical String for this request should have been
'DELETE
/_search/scroll

... remaining request info...

### How can one reproduce the bug?
How can one reproduce the bug?

Create an OpenSearchClient following instructions from here (https://opensearch.org/docs/2.7/clients/java/#connecting-to-amazon-opensearch-service) and then attempt to use the client's `clearScroll` API.

I believe this is because it is a DELETE request with a body. The `search/scroll` method works as intended. Sending the request results in a sigv4 error

### What is the expected behavior?
The `clearScroll` API call does not have a signature mismatch

### What is your host/environment?
Ubuntu 22.04
`implementation 'org.opensearch.client:opensearch-java:2.8.0'`

### Do you have any screenshots?
_If applicable, add screenshots to help explain your problem._

### Do you have any additional context?
_Add any other context about the problem._

### PR DESCRIPTION
**Pull request by Xtansia**
### Description
The AWS SDK's ApacheHttpClient implementation does not send the request body on DELETE or GET requests, https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/impl/ApacheHttpRequestFactory.java#L118-L137.

Additionally moves to the supported `AwsV4HttpSigner` as `Aws4Signer` is now deprecated: https://github.com/aws/aws-sdk-java-v2/blob/88abec27e7d5d35b21545c7e05875a7cc3d0f46e/core/auth/src/main/java/software/amazon/awssdk/auth/signer/Aws4Signer.java

### Issues Resolved
Relates to #712 
Relates to #521 
Fixes #503 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: java-client/build.gradle.kts [hunk 1]
```diff
@@ -210,21 +209,25 @@ dependencies {
...
-    "awsSdk2SupportCompileOnly"("software.amazon.awssdk","auth","[2.15,3.0)")
-    testImplementation("software.amazon.awssdk","sdk-core","[2.15,3.0)")
-    testImplementation("software.amazon.awssdk","auth","[2.15,3.0)")
-    testImplementation("software.amazon.awssdk","aws-crt-client","[2.15,3.0)")
-    testImplementation("software.amazon.awssdk","apache-client","[2.15,3.0)")
-    testImplementation("software.amazon.awssdk","sts","[2.15,3.0)")
+    "awsSdk2SupportCompileOnly"("software.amazon.awssdk", "sdk-core", "[2.21,3.0)")
+    "awsSdk2SupportCompileOnly"("software.amazon.awssdk", "auth", "[2.21,3.0)")
+    "awsSdk2SupportCompileOnly"("software.amazon.awssdk", "http-auth-aws", "[2.21,3.0)")
+    testImplementation("software.amazon.awssdk", "sdk-core", "[2.21,3.0)")
```
Comment (reta) [inline java-client/build.gradle.kts:220]: I believe we should be using `awsSdk2SupportTestImplementation` here:

```suggestion
    awsSdk2SupportTestImplementation("software.amazon.awssdk", "sdk-core", "[2.21,3.0)")
```

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java [hunk 2]
```diff
@@ -85,6 +87,7 @@ public class AwsSdk2Transport implements OpenSearchTransport {
     private final String signingServiceName;
     private final Region signingRegion;
     private final JsonpMapper defaultMapper;
+    @Nonnull
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java:90]: No particular take on that, but I believe all this fields are should not be null.

File: java-client/build.gradle.kts [hunk 3]
```diff
@@ -348,40 +351,43 @@ publishing {
...
+    configurations[java11.implementationConfigurationName].extendsFrom(configurations.testImplementation.get())
+    configurations[java11.runtimeOnlyConfigurationName].extendsFrom(configurations.testRuntimeOnly.get())
+
+    dependencies {
+        testImplementation("org.opensearch.test", "framework", opensearchVersion) {
+            exclude(group = "org.hamcrest")
+        }
+
+        // Apache 2.0
+        testImplementation("org.wiremock", "wiremock", "3.9.2")
```
Comment (reta) [inline java-client/build.gradle.kts:371]: Nothing against `wiremock`, but we already have all needed pieces:
 - `HttpServer` that comes from HttpCore 5 (example https://github.com/apache/httpcomponents-core/blob/master/httpcore5-reactive/src/test/java/org/apache/hc/core5/reactive/examples/ReactiveFullDuplexServerExample.java)
 - since we run test with JDK-11, we could use `com.sun.net.httpserver.HttpServer` that comes with standard library (see please https://docs.oracle.com/en/java/javase/11/docs/api/jdk.httpserver/com/sun/net/httpserver/HttpServer.html)

File: java-client/build.gradle.kts (hunk 1 above)
Comment (Xtansia) [reply java-client/build.gradle.kts:243]: Such a configuration doesn't exist, the `awsSdk2Support` feature/variant doesn't affect the test source set

File: java-client/build.gradle.kts (hunk 3 above)
Comment (Xtansia) [reply java-client/build.gradle.kts]: Have dropped wiremock and re-implemented the necessary features

**Review by dblock — DISMISSED** [review]
Looks good. A lot going on that's somewhat unrelated to the signing fix, you might want to extract it first, but up to you.

File: CHANGELOG.md [hunk 4]
```diff
@@ -51,9 +48,21 @@ This section is for maintaining a changelog for all breaking changes for the cli
 ### Removed
 
 ### Fixed
+- Fixed AWS SigV4 on delete requests when using AWS SDK's Apache client ([#1256](https://github.com/opensearch-project/opensearch-java/pull/1256))
 
 ### Security
 
+## [2.17.0] - 11/05/2024
```
Comment (reta) [inline CHANGELOG.md:55]: Nit, seems to be unrelated changes?

File: java-client/src/test/java/org/opensearch/client/transport/util/GeneratedCertificateSSLContext.java [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+import sun.security.x509.KeyUsageExtension;
+import sun.security.x509.SerialNumber;
+import sun.security.x509.SubjectAlternativeNameExtension;
+import sun.security.x509.SubjectKeyIdentifierExtension;
+import sun.security.x509.X500Name;
+import sun.security.x509.X509CertImpl;
+import sun.security.x509.X509CertInfo;
+
+@SuppressWarnings("sunapi")
+public class GeneratedCertificateSSLContext {
```
Comment (reta) [inline java-client/src/test/java/org/opensearch/client/transport/util/GeneratedCertificateSSLContext.java:64]: I think it would be much simpler to pre-generate certs, no? (see please https://github.com/opensearch-project/OpenSearch/tree/main/plugins/transport-reactor-netty4/src/test/resources).

Comment (reta) [reply java-client/src/test/java/org/opensearch/client/transport/util/GeneratedCertificateSSLContext.java:64]: (at worst, the security plugin uses Bouncycastle to generate those https://github.com/opensearch-project/security/blob/main/src/test/java/org/opensearch/security/ssl/CertificatesRule.java)

Comment (Xtansia) [reply java-client/src/test/java/org/opensearch/client/transport/util/GeneratedCertificateSSLContext.java:64]: Pre-generating and committing the certs unfortunately leads to a lot of false positive security alerts. I'd initially avoided using BC seeing as the changes were with the aim to avoid pulling in another dependency. But can swap over to BC

File: java-client/src/test/java/org/opensearch/client/transport/aws/AwsSdk2TransportTests.java [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        NETTY_NIO_ASYNC
+    }
+
+    private static final String[] TEST_SERVICE_NAMES = { "aoss", "es", "arbitrary" };
+    private static final Region TEST_REGION = Region.AP_SOUTHEAST_2;
+    private static final String TEST_INDEX = "sample-index1";
+    private static final SSLContext SSL_CONTEXT;
+    private static final TrustManager[] CLIENT_TRUST_MANAGERS;
+
+    static {
```
Comment (reta) [inline java-client/src/test/java/org/opensearch/client/transport/aws/AwsSdk2TransportTests.java:89]: Why not to use @BeforeClass here?

File: java-client/src/test/java/org/opensearch/client/transport/util/SelfSignedCertificateAuthority.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter;
+import org.bouncycastle.cert.jcajce.JcaX509ExtensionUtils;
+import org.bouncycastle.cert.jcajce.JcaX509v3CertificateBuilder;
+import org.bouncycastle.crypto.CryptoServicesRegistrar;
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+import org.bouncycastle.operator.ContentSigner;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+
+public class SelfSignedCertificateAuthority {
```
Comment (reta) [inline java-client/src/test/java/org/opensearch/client/transport/util/SelfSignedCertificateAuthority.java:43]: :+1: TY

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java [hunk 8]
```diff
@@ -315,46 +312,57 @@ private <RequestT> SdkHttpFullRequest prepareRequest(
         } catch (URISyntaxException e) {
             throw new IllegalArgumentException("Invalid request URI: " + url.toString());
         }
+
+        ContentStreamProvider bodyProvider = body != null ? ContentStreamProvider.fromInputStreamSupplier(body::getInputStream) : null;
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java:316]: So the `body` is `OpenSearchRequestBodyBuffer` and is always backed by buffer, could we use `ContentStreamProvider::fromByteArrayUnsafe` (zero copy) instead of creating a wrapping input stream over the array?

**Comment by Xtansia** [comment]
Unfortunately on further digging and expanding the tests a bit, AWS SDK's ApacheHttpClient does not send the request body at all for DELETE or GET requests, https://github.com/aws/aws-sdk-java-v2/blob/master/http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/impl/ApacheHttpRequestFactory.java#L118-L137

So this doesn't actually fix anything, as such I've pivoted to pre-emptively throwing an exception as the request would fail signature validation anyway and gives a clear reasoning.

**Comment by reta** [comment]
> @Xtansia Probably worth opening an issue in that SDK, AFAIK nothing says `GET` or `DELETE` can't have a body.

@dblock indeed the spec [1] is not strict on that but the general recommendation (like [2]) is that GET should not have a body, DELETE though may [3]. Fe, https://www.rfc-editor.org/rfc/rfc9110.html#name-get says (similar for DELETE):


> Although request message framing is independent of the method used, content received in a GET request has no generally defined semantics, cannot alter the meaning or target of the request, and might lead some implementations to reject the request and close the connection because of its potential as a request smuggling attack ([Section 11.2](https://www.rfc-editor.org/rfc/rfc9112#section-11.2) of [[HTTP/1.1](https://www.rfc-editor.org/rfc/rfc9110.html#HTTP11)]).  A client SHOULD NOT generate content in a GET request unless it is made directly to an origin server that has previously indicated, in or out of band, that such a request has a purpose and will be adequately supported. An origin server SHOULD NOT rely on private agreements to receive content, since participants in HTTP communication are often unaware of intermediaries along the request chain.

[1] https://www.rfc-editor.org/rfc/rfc9110.html#name-changes-from-rfc-7231
[2] https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/GET
[3] https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/DELETE

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java [hunk 9]
```diff
@@ -386,14 +424,14 @@ private void applyOptionsHeaders(SdkHttpFullRequest.Builder builder, TransportOp
...
+        SdkHttpRequest httpRequest,
+        OpenSearchRequestBodyBuffer requestBody,
         Endpoint<?, ResponseT, ?> endpoint,
         TransportOptions options
     ) throws IOException {
-
         HttpExecuteRequest.Builder executeRequest = HttpExecuteRequest.builder().request(httpRequest);
-        if (httpRequest.contentStreamProvider().isPresent()) {
-            executeRequest.contentStreamProvider(httpRequest.contentStreamProvider().get());
+        if (requestBody != null) {
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java:433]: Why we cannot reuse the `contentStreamProvider` (as before) but have to pass `requestBody` and create a new `ContentStreamProvider` ?

File: java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java [hunk 10]
```diff
@@ -118,6 +122,10 @@ public AwsSdk2Transport(
      * @param options Options that apply to all requests. Can be null. Create with
      *                {@link AwsSdk2TransportOptions#builder()} and use these to specify non-default credentials,
      *                compression options, etc.
+     *
+     * @implNote Using {@code software.amazon.awssdk.http.apache.ApacheHttpClient} is discouraged as it does not support request bodies on GET or DELETE requests.
```
Comment (reta) [inline java-client/src/main/java/org/opensearch/client/transport/aws/AwsSdk2Transport.java:126]: :+1:

**Comment by Xtansia** [comment]
I might just try PR'ing a fix into aws-sdk seeing as it's relatively simple change, and hopefully mean more likely to actually get fixed.

**Comment by Xtansia** [comment]
Created https://github.com/aws/aws-sdk-java-v2/pull/5704

**Comment by dblock** [comment]
> Created [aws/aws-sdk-java-v2#5704](https://github.com/aws/aws-sdk-java-v2/pull/5704)

Good work 👏

**Review by reta — APPROVED** [review]

**Comment by Xtansia** [comment]
@houssain-barouni Yes, I have created #1333
